# J1S1 — Setup & GitHub Copilot
**Formation BankRisk Intelligence Platform · Jour 1 · Session 1 · 09h00 – 10h30**

> **Repo formateur :** `alexdiby/bankrisk-credit`  
> **Ton repo :** créé depuis GitHub.com (pas un fork — repo personnel)  
> **Dataset :** `credit_risk_dataset.csv` (32 581 lignes · 12 colonnes · CC0)  
> **Taux de défaut de référence : 21,8 %**

---

## 🎯 Objectifs

- Vérifier que l'environnement installé via le guide HTML est opérationnel
- Créer la structure standard du projet depuis le terminal
- Configurer `requirements.txt` et `.gitignore`
- Pratiquer les 3 modes GitHub Copilot (complétion, inline `Ctrl+I`, chat latéral `Ctrl+Alt+I`)
- Charger le dataset fil rouge et valider le taux de défaut : **21,8 %**
- Générer le `README.md` et faire un premier commit

---

> ⚠️ **Prérequis** : avoir complété toutes les étapes du guide HTML d'installation.  
> Le dépôt doit être cloné, le venv actif, et VS Code ouvert dans `workspace/bankrisk-credit`.  
> Si ce n'est pas le cas, reprendre le guide à partir de la **Phase 0**.


---

## 🔗 Continuité avec le guide d'installation

**Ce que vous avez fait dans le guide HTML :**

| Phase | Étape | Statut |
|-------|-------|--------|
| 0 | Dossier `workspace/` créé, `code .` fonctionne | ✅ |
| 1 | Git installé et configuré (`user.name`, `user.email`) | ✅ |
| 2 | Compte GitHub créé + Copilot Free activé | ✅ |
| 3 | Dépôt `bankrisk-credit` créé sur GitHub + cloné dans `workspace/` | ✅ |
| 4 | Python 3.10.14 installé | ✅ |
| 5 | venv `.venv` créé et activé — packages installés | ✅ |
| 6 | VS Code installé + connecté à GitHub | ✅ |
| 7 | Extensions Python, Jupyter, Copilot, Copilot Chat installées | ✅ |
| 8 | Copilot actif (icône verte en bas de VS Code) | ✅ |
| 9 | Vérification finale OK — `pyarrow` importable | ✅ |

**Ce que vous allez faire dans ce notebook :**

1. Vérifier l'environnement Python depuis le notebook (Bloc 1)
2. Créer la structure de projet depuis le **terminal** (Bloc 2)
3. Placer le dataset dans `data/raw/` (Bloc 3)
4. Générer `requirements.txt` et `.gitignore` (Bloc 4)
5. Pratiquer GitHub Copilot — 5 exercices (Bloc 5)
6. Charger et explorer le dataset fil rouge (Bloc 6)
7. Générer le `README.md` et faire un commit (Bloc 7)


---

## Bloc 1 — Vérification de l'environnement

> **Avant d'exécuter :** vérifier que le kernel affiché en haut à droite de VS Code est  
> **`.venv (Python 3.10.x)`** — pas le Python système.  
> Si ce n'est pas le cas : `Ctrl+Shift+P` → `Python: Select Interpreter` → `.venv`


In [ ]:
# ── Cellule 1 : ROOT + imports + vérification environnement ─────────────────
# Cette cellule doit être exécutée EN PREMIER à chaque session
import sys
from pathlib import Path
import os

# Détection de la racine du projet
ROOT = Path.cwd()
# Si le notebook est dans notebooks/, remonter d'un niveau
if ROOT.name == "notebooks":
    ROOT = ROOT.parent

os.chdir(ROOT)
print(f"📁 Racine du projet : {ROOT}")
print(f"📂 Répertoire courant : {Path.cwd()}")
print()

# Versions Python
print(f"Python  : {sys.version.split()[0]}")
print(f"Chemin  : {sys.executable}")
print()

# Détection de l'environnement
if '.venv' in sys.executable:
    print("✅ Virtualenv .venv actif — VS Code")
elif 'google.colab' in sys.modules:
    print("✅ Google Colab détecté")
else:
    print("⚠️  Environnement non reconnu")
    print(f"   Interpréteur : {sys.executable}")
    print("   → Ctrl+Shift+P → Python: Select Interpreter → .venv")


In [ ]:
# ── Cellule 2 : Vérification des packages ───────────────────────────────────
import importlib

packages = {
    "pandas":    ("pd",  "Manipulation de données — J1S2"),
    "numpy":     ("np",  "Calcul vectorisé — J1S2"),
    "plotly":    (None,  "Visualisation — J1S3"),
    "sklearn":   (None,  "Machine learning — J2S2"),
    "pyarrow":   (None,  "Format Parquet — J1S2"),
    "mlflow":    (None,  "Experiment tracking — J2S3"),
}

print("Packages installés :")
all_ok = True
for pkg, (alias, usage) in packages.items():
    try:
        mod = importlib.import_module(pkg)
        ver = getattr(mod, "__version__", "?")
        print(f"  ✅ {pkg:<14} {ver:<10} — {usage}")
    except ImportError:
        print(f"  ❌ {pkg:<14} {'MANQUANT':<10} — {usage}")
        all_ok = False

print()
if all_ok:
    print("✅ Tous les packages sont disponibles — environnement opérationnel")
else:
    print("⚠️  Packages manquants — exécuter dans le terminal :")
    print("   pip install -r requirements.txt")


### 💡 Résolution des erreurs courantes

**`ModuleNotFoundError` sous VS Code**  
→ Le mauvais kernel est sélectionné. Cliquer **Select Kernel** (haut droite) → `.venv (Python 3.10.x)`

**`ModuleNotFoundError` après sélection du venv**  
→ Le venv n'a pas encore les packages. Ouvrir le terminal intégré (`Ctrl+\``) et exécuter :
```bash
pip install -r requirements.txt
```


---

## Bloc 2 — Structure du projet

La structure ci-dessous sera identique pour **toutes les sessions de la formation**.

```
bankrisk-credit/
├── data/
│   ├── raw/            ← Dataset original — jamais modifié, non versionné
│   └── processed/      ← Données transformées (Parquet, CSV)
├── notebooks/          ← Notebooks Jupyter par session (J1S1, J1S2…)
├── src/                ← Scripts Python réutilisables
├── tests/              ← Tests unitaires (pytest)
├── requirements.txt    ← Dépendances Python
└── .gitignore          ← Fichiers exclus du contrôle de version
```

> **Règle absolue :** `data/raw/` n'est **jamais modifié** après dépôt initial.  
> Toutes les transformations produisent des fichiers dans `data/processed/`.

---

### 🖥️ Créer la structure depuis le terminal

Ouvrir le terminal intégré VS Code (`Ctrl+\`` ) et exécuter les commandes suivantes :

```bash
# Se placer à la racine du projet (normalement déjà là)
cd workspace/bankrisk-credit

# Créer tous les dossiers d'un coup
mkdir -p data/raw data/processed notebooks src tests

# Créer les .gitkeep pour que Git versionne les dossiers vides
touch data/raw/.gitkeep
touch data/processed/.gitkeep
touch notebooks/.gitkeep
touch src/.gitkeep
touch tests/.gitkeep
```

> **Pourquoi les `.gitkeep` ?**  
> Git ne traque pas les dossiers vides. Un fichier `.gitkeep` (vide, par convention) permet  
> de committer la structure sans inclure les données.


In [ ]:
# ── Vérification de la structure créée ──────────────────────────────────────
# Cette cellule vérifie que les commandes bash ci-dessus ont bien été exécutées

from pathlib import Path

expected = [
    "data/raw",
    "data/processed",
    "notebooks",
    "src",
    "tests",
]

print("Vérification de la structure du projet :")
print(f"Racine : {ROOT}")
print()

all_ok = True
for folder in expected:
    path = ROOT / folder
    gitkeep = path / ".gitkeep"
    if path.exists():
        gk = "✅ .gitkeep présent" if gitkeep.exists() else "⚠️  .gitkeep manquant"
        print(f"  ✅ {folder}/   {gk}")
    else:
        print(f"  ❌ {folder}/   → dossier manquant")
        print(f"     Exécuter dans le terminal : mkdir -p {folder} && touch {folder}/.gitkeep")
        all_ok = False

print()
if all_ok:
    print("✅ Structure complète — prêt pour la suite")
else:
    print("⚠️  Reprendre les commandes bash du Bloc 2 dans le terminal")


---

## Bloc 3 — Placer le dataset

Le dataset `credit_risk_dataset.csv` est le **fil rouge de toute la formation**.  
Il doit être placé manuellement dans `data/raw/` — il n'est **jamais commité** dans le repo.

---

### 🖥️ Télécharger et placer le dataset

**Option A — Téléchargement depuis Kaggle**

1. Aller sur : https://www.kaggle.com/datasets/laotse/credit-risk-dataset
2. Cliquer **Download** → décompresser le `.zip`
3. Copier `credit_risk_dataset.csv` dans `workspace/bankrisk-credit/data/raw/`

**Option B — Copier depuis le dossier de téléchargements (terminal)**

```bash
# Windows (Git Bash)
cp ~/Downloads/credit_risk_dataset.csv data/raw/

# macOS / Linux
cp ~/Downloads/credit_risk_dataset.csv data/raw/
```

**Option C — Lien direct fourni par le formateur**

Le formateur met le fichier à disposition via le repo `alexdiby/bankrisk-credit`.

```bash
# Télécharger avec curl (si disponible)
curl -L -o data/raw/credit_risk_dataset.csv \
  "https://raw.githubusercontent.com/alexdiby/bankrisk-credit/main/data/raw/credit_risk_dataset.csv"
```


In [ ]:
# ── Vérification de la présence du dataset ──────────────────────────────────
from pathlib import Path

csv_path = ROOT / "data" / "raw" / "credit_risk_dataset.csv"

if csv_path.exists():
    size_mb = csv_path.stat().st_size / (1024 * 1024)
    print(f"✅ Dataset trouvé : {csv_path}")
    print(f"   Taille : {size_mb:.2f} Mo")
else:
    print(f"❌ Dataset non trouvé : {csv_path}")
    print()
    print("   → Télécharger credit_risk_dataset.csv et le copier dans data/raw/")
    print("   → Voir les instructions du Bloc 3 ci-dessus")


---

## Bloc 4 — `requirements.txt` et `.gitignore`

### 🖥️ Créer `requirements.txt` depuis le terminal

Le fichier liste toutes les dépendances Python de la formation.  
Exécuter dans le terminal intégré VS Code :

```bash
# Générer le requirements.txt à partir des packages installés dans le venv
pip freeze > requirements.txt
```

> **Résultat attendu :** un fichier `requirements.txt` à la racine du projet avec toutes les versions exactes.

---

### 🖥️ Créer le `.gitignore` depuis le terminal

```bash
# Sur macOS / Linux / Git Bash (Windows)
cat > .gitignore << 'EOF'
# ── Données — jamais commitées ──────────────────────────────
data/raw/*
data/processed/*
!data/raw/.gitkeep
!data/processed/.gitkeep
*.csv
*.parquet
*.xlsx

# ── Modèles et MLflow ────────────────────────────────────────
*.pkl
*.joblib
mlruns/

# ── Python ──────────────────────────────────────────────────
__pycache__/
*.py[cod]
*.egg-info/
dist/
build/
.venv/
env/
venv/

# ── Jupyter ─────────────────────────────────────────────────
.ipynb_checkpoints/

# ── Secrets ─────────────────────────────────────────────────
.env
.env.*
secrets.yaml

# ── OS ──────────────────────────────────────────────────────
.DS_Store
Thumbs.db
EOF
```

> **Sur Windows (PowerShell)**, préférer créer le fichier directement dans VS Code :  
> `Ctrl+N` → coller le contenu → `Ctrl+Shift+S` → nommer `.gitignore` à la racine du projet.


In [ ]:
# ── Vérification requirements.txt et .gitignore ─────────────────────────────
from pathlib import Path

req_path = ROOT / "requirements.txt"
gi_path  = ROOT / ".gitignore"

# requirements.txt
if req_path.exists():
    lines = req_path.read_text(encoding="utf-8").splitlines()
    lines = [l for l in lines if l.strip() and not l.startswith("#")]
    print(f"✅ requirements.txt — {len(lines)} packages listés")
    critical = ["pandas", "numpy", "plotly", "scikit-learn", "pyarrow", "mlflow"]
    for pkg in critical:
        found = any(pkg.lower() in l.lower() for l in lines)
        status = "✅" if found else "⚠️  manquant"
        print(f"   {status}  {pkg}")
else:
    print("❌ requirements.txt non trouvé")
    print("   → Exécuter dans le terminal : pip freeze > requirements.txt")

print()

# .gitignore
if gi_path.exists():
    content = gi_path.read_text(encoding="utf-8")
    checks = {
        "data/raw/*":           "Données raw ignorées",
        "!data/raw/.gitkeep":   "Structure raw conservée",
        "*.parquet":            "Parquet ignoré",
        ".venv/":               "venv ignoré",
        ".env":                 "Secrets ignorés",
    }
    print(".gitignore — vérification des règles clés :")
    for rule, desc in checks.items():
        status = "✅" if rule in content else "⚠️  absent"
        print(f"   {status}  {rule:<30} {desc}")
else:
    print("❌ .gitignore non trouvé")
    print("   → Voir les commandes bash du Bloc 4")


---

## Bloc 5 — GitHub Copilot : 5 exercices progressifs

### Rappel des raccourcis

| Raccourci | Mode | Action |
|-----------|------|--------|
| Taper + attendre 2 sec | **Complétion inline** | Copilot propose en gris → `Tab` pour accepter |
| `Tab` | — | Accepter la suggestion |
| `Échap` | — | Refuser |
| `Alt+]` / `Alt+[` | — | Suggestion suivante / précédente |
| `Ctrl+I` | **Chat inline** | Modifier / expliquer le code sélectionné |
| `Ctrl+Alt+I` | **Chat latéral** | Questions complexes, génération de fonctions |

> **Règle d'or :** comprendre avant d'accepter — Copilot peut générer du code incorrect.


---

### 🧪 Exercice 1 — Technique du commentaire-intention

**Principe :** écrire ce qu'on veut faire en commentaire → attendre 2 secondes → `Tab`.

**Consigne :** taper **uniquement les commentaires** ci-dessous, un par un.  
Laisser Copilot compléter chaque ligne de code avant de passer au commentaire suivant.

> ⚠️ Ne pas copier-coller le code — l'objectif est de voir Copilot le générer en temps réel.


In [ ]:
import pandas as pd
import numpy as np

# Charger le dataset credit risk depuis le fichier CSV
df = pd.read_csv(ROOT / "data/raw/credit_risk_dataset.csv")

# Afficher le nombre de lignes et colonnes
print(f"Shape : {df.shape}")

# Calculer le taux de défaut moyen
taux = df["loan_status"].mean()
print(f"Taux de défaut : {taux:.1%}")

# Afficher les 3 premières lignes
df.head(3)


---

### 🧪 Exercice 2 — Chat inline `Ctrl+I` : comprendre du code

**Principe :** sélectionner du code existant → `Ctrl+I` → poser une question en langage naturel.

**Consigne :**
1. Sélectionner **tout le code** de la cellule ci-dessous
2. Appuyer sur `Ctrl+I`
3. Taper : `explique ce que fait ce code ligne par ligne`
4. Lire la réponse de Copilot dans la fenêtre inline

> **Objectif :** utiliser Copilot comme outil de compréhension, pas seulement de génération.


In [ ]:
# Code à analyser avec Copilot (Ctrl+I → "explique ce que fait ce code")
import pandas as pd

df = pd.read_csv(ROOT / "data/raw/credit_risk_dataset.csv")

risk_by_grade = (
    df.groupby("loan_grade")["loan_status"]
    .agg(["mean", "count"])
    .rename(columns={"mean": "taux_defaut", "count": "nb_prets"})
    .sort_values("taux_defaut", ascending=False)
    .round(3)
)
print(risk_by_grade)


---

### 🧪 Exercice 3 — Chat inline `Ctrl+I` : corriger une erreur

**Principe :** le code ci-dessous contient **3 erreurs volontaires**.  
Utiliser Copilot pour les identifier et les corriger.

**Consigne :**
1. Sélectionner tout le code
2. `Ctrl+I` → taper : `ce code contient des erreurs, corrige-les`
3. Comparer la correction proposée avec le code original
4. Accepter si la correction est correcte — **vérifier avant d'accepter**

> Erreurs cachées : une faute de colonne, un mauvais type, une logique inversée.


In [ ]:
import pandas as pd

df = pd.read_csv(ROOT / "data/raw/credit_risk_dataset.csv")

# Erreur 1 : mauvais nom de colonne
revenu_moyen = df["person_revenues"].mean()  # ← colonne inexistante

# Erreur 2 : mauvais type pour la comparaison
hauts_revenus = df[df["person_income"] > "50000"]  # ← comparaison str/int

# Erreur 3 : logique inversée
taux_sains = df["loan_status"].sum() / len(df)  # ← calcule le taux de défauts, pas de sains

print(f"Revenu moyen : {revenu_moyen:.0f}")
print(f"Hauts revenus : {len(hauts_revenus)} lignes")
print(f"Taux de prêts sains : {taux_sains:.1%}")


---

### 🧪 Exercice 4 — Chat inline `Ctrl+I` : refactorer du code

**Principe :** améliorer du code fonctionnel mais peu lisible.

**Consigne :**
1. Sélectionner tout le code
2. `Ctrl+I` → taper : `refactorise ce code pour le rendre plus lisible et ajoute des commentaires`
3. Observer comment Copilot restructure sans changer le comportement


In [ ]:
import pandas as pd
df = pd.read_csv(ROOT / "data/raw/credit_risk_dataset.csv")
# Code à refactorer — fonctionnel mais illisible
result = df[df["loan_status"]==1].groupby("loan_grade")["loan_amnt"].mean().reset_index().rename(columns={"loan_amnt":"montant_moyen_defaut"}).sort_values("montant_moyen_defaut",ascending=False).round(0)
print(result)


---

### 🧪 Exercice 5 — Chat latéral `Ctrl+Alt+I` : générer une fonction complète

**Principe :** le chat latéral permet de décrire une fonction complexe en langage naturel.

**Consigne :**
1. Ouvrir le chat latéral : `Ctrl+Alt+I`
2. Coller ce prompt exact :

```
Écris une fonction Python appelée `calculer_taux_defaut` qui :
- prend en paramètre un DataFrame pandas et un nom de colonne de segmentation
- retourne un DataFrame avec le taux de défaut (loan_status) par segment
- trié par taux décroissant
- inclut le nombre de prêts par segment
- ajoute une docstring complète en français
```

3. Copier le code généré dans la cellule ci-dessous
4. Tester la fonction sur `loan_grade` puis sur `person_home_ownership`


In [3]:
import pandas as pd


def calculer_taux_defaut(df: pd.DataFrame, colonne_segmentation: str) -> pd.DataFrame:
    """
    Calcule le taux de défaut par segment à partir d'un DataFrame.

    Cette fonction agrège la colonne `loan_status` par segment défini par
    `colonne_segmentation` et retourne un DataFrame contenant :
    - le taux de défaut moyen par segment,
    - le nombre total de prêts dans chaque segment,
    - les segments triés par taux de défaut décroissant.

    Paramètres
    ----------
    df : pandas.DataFrame
        DataFrame contenant au moins les colonnes `loan_status` et la colonne
        de segmentation fournie en paramètre.
    colonne_segmentation : str
        Nom de la colonne utilisée pour créer les segments.

    Retour
    ------
    pandas.DataFrame
        DataFrame avec les colonnes `colonne_segmentation`, `taux_defaut` et
        `nb_prets`, trié par `taux_defaut` décroissant.

    Raises
    ------
    KeyError
        Si `loan_status` ou la colonne de segmentation est absente.
    ValueError
        Si le DataFrame est vide.
    """
    if df.empty:
        raise ValueError("Le DataFrame ne doit pas être vide.")

    if "loan_status" not in df.columns:
        raise KeyError("La colonne 'loan_status' est absente du DataFrame.")

    if colonne_segmentation not in df.columns:
        raise KeyError(f"La colonne '{colonne_segmentation}' est absente du DataFrame.")

    resultat = (
        df.groupby(colonne_segmentation, dropna=False)["loan_status"]
        .agg(["mean", "count"])
        .rename(columns={"mean": "taux_defaut", "count": "nb_prets"})
        .sort_values("taux_defaut", ascending=False)
        .reset_index()
    )

    resultat.index.name = colonne_segmentation
    return resultat


# ── Test de la fonction ──────────────────────────────────────────────────────
df = pd.read_csv("C:/Users/HP/Desktop/Project Formation/credit-risk-formation/data/raw/credit_risk_dataset.csv")

# Test 1 : segmentation par grade
print(calculer_taux_defaut(df, "loan_grade").head())

# Test 2 : segmentation par type de logement
print(calculer_taux_defaut(df, "person_home_ownership").head())


           loan_grade  taux_defaut  nb_prets
loan_grade                                  
0                   G     0.984375        64
1                   F     0.705394       241
2                   E     0.644191       964
3                   D     0.590458      3626
4                   C     0.207340      6458
                      person_home_ownership  taux_defaut  nb_prets
person_home_ownership                                             
0                                      RENT     0.315700     16446
1                                     OTHER     0.308411       107
2                                  MORTGAGE     0.125707     13444
3                                       OWN     0.074690      2584


---

## Bloc 6 — Charger et explorer le dataset fil rouge

| Propriété | Valeur |
|-----------|--------|
| **Lignes** | 32 581 |
| **Colonnes** | 12 |
| **Cible** | `loan_status` — 0 = sain, 1 = défaut |
| **Taux de défaut** | **21,8 %** |
| **Valeurs manquantes** | `loan_int_rate` : 3 116 (9,6 %) · `person_emp_length` : 895 (2,7 %) |
| **Outliers** | `person_age` max 144 · `person_emp_length` max 123 |


In [5]:
import pandas as pd
import numpy as np

# ── Chargement ──────────────────────────────────────────────────────────────
df = pd.read_csv("C:/Users/HP/Desktop/Project Formation/credit-risk-formation/data/raw/credit_risk_dataset.csv")

print(f"Lignes   : {df.shape[0]:,}")
print(f"Colonnes : {df.shape[1]}")
print()
print("Colonnes et types :")
for col in df.columns:
    print(f"  {col:<35} {str(df[col].dtype)}")


Lignes   : 32,581
Colonnes : 12

Colonnes et types :
  person_age                          int64
  person_income                       int64
  person_home_ownership               object
  person_emp_length                   float64
  loan_intent                         object
  loan_grade                          object
  loan_amnt                           int64
  loan_int_rate                       float64
  loan_status                         int64
  loan_percent_income                 float64
  cb_person_default_on_file           object
  cb_person_cred_hist_length          int64


In [6]:
# ── Aperçu des 5 premières lignes ───────────────────────────────────────────
df.head(5)


,person_age,person_income,person_home_ownership,person_emp_length,loan_intent,loan_grade,loan_amnt,loan_int_rate,loan_status,loan_percent_income,cb_person_default_on_file,cb_person_cred_hist_length
0,22,59000,RENT,123.0,PERSONAL,D,35000,16.02,1,0.59,Y,3
1,21,9600,OWN,5.0,EDUCATION,B,1000,11.14,0,0.10,N,2
2,25,9600,MORTGAGE,1.0,MEDICAL,C,5500,12.87,1,0.57,N,3
3,23,65500,RENT,4.0,MEDICAL,C,35000,15.23,1,0.53,N,2
4,24,54400,RENT,8.0,MEDICAL,C,35000,14.27,1,0.55,Y,4


In [7]:
# ── Statistiques descriptives ───────────────────────────────────────────────
df.describe().round(2)


,person_age,person_income,person_emp_length,loan_amnt,loan_int_rate,loan_status,loan_percent_income,cb_person_cred_hist_length
count,32581.00,32581.00,31686.00,32581.00,29465.00,32581.00,32581.00,32581.00
mean,27.73,66074.85,4.79,9589.37,11.01,0.22,0.17,5.80
std,6.35,61983.12,4.14,6322.09,3.24,0.41,0.11,4.06
min,20.00,4000.00,0.00,500.00,5.42,0.00,0.00,2.00
25%,23.00,38500.00,2.00,5000.00,7.90,0.00,0.09,3.00
50%,26.00,55000.00,4.00,8000.00,10.99,0.00,0.15,4.00
75%,30.00,79200.00,7.00,12200.00,13.47,0.00,0.23,8.00
max,144.00,6000000.00,123.00,35000.00,23.22,1.00,0.83,30.00


In [8]:
# ── Valeurs manquantes ──────────────────────────────────────────────────────
missing     = df.isnull().sum()
missing_pct = (missing / len(df) * 100).round(2)

missing_df = pd.DataFrame({
    "Manquants":      missing,
    "Pourcentage (%)": missing_pct
}).sort_values("Manquants", ascending=False)

print("Valeurs manquantes par colonne :")
print(missing_df[missing_df["Manquants"] > 0])
print()
print("⚠️  Ces valeurs manquantes seront traitées en J1S4 (Feature Engineering)")


Valeurs manquantes par colonne :
                   Manquants  Pourcentage (%)
loan_int_rate           3116             9.56
person_emp_length        895             2.75

⚠️  Ces valeurs manquantes seront traitées en J1S4 (Feature Engineering)


In [9]:
# ── Taux de défaut — chiffre de référence de toute la formation ─────────────
print("=" * 50)
print("Distribution de la variable cible : loan_status")
print("=" * 50)
print()
print(df["loan_status"].value_counts())
print()
print(df["loan_status"].value_counts(normalize=True).round(3))
print()
taux_defaut = df["loan_status"].mean()
print(f"Taux de défaut : {taux_defaut:.1%}")
print()
print("→ À retenir pour toute la formation : 21,8 %")
print("  0 = prêt sain  ·  1 = défaut de paiement")


Distribution de la variable cible : loan_status

loan_status
0    25473
1     7108
Name: count, dtype: int64

loan_status
0    0.782
1    0.218
Name: proportion, dtype: float64

Taux de défaut : 21.8%

→ À retenir pour toute la formation : 21,8 %
  0 = prêt sain  ·  1 = défaut de paiement


---

## Bloc 7 — `README.md` et premier commit

### Étape 1 : Générer le README avec Copilot (chat latéral `Ctrl+Alt+I`)

Ouvrir le chat latéral et coller ce prompt :

```
Génère un README.md professionnel en français pour un projet Data Science avec cette structure :

Nom du projet : Credit Risk Analysis — BankRisk Intelligence Platform
Description : Analyse du risque de crédit sur le dataset Kaggle Credit Risk
(32 581 prêts, cible binaire loan_status : 0=sain, 1=défaut). Taux de défaut : 21,8 %.

Stack : Python 3.10, Pandas, NumPy, Plotly, Scikit-learn, MLflow, DagsHub,
Hugging Face, PySpark, Delta Lake, Streamlit, GitHub Actions.

Sections : badges, description, structure du repo (ASCII), installation, dataset, stack, roadmap 3 jours, licence CC0.
```

### Étape 2 : Écrire le README depuis Python

La cellule ci-dessous génère un README de base. Remplacer `TON_USERNAME` avant d'exécuter.


In [ ]:
# ── Génération du README.md ──────────────────────────────────────────────────
# ⚠️ Remplacer TON_USERNAME par ton identifiant GitHub
GITHUB_USERNAME = "TON_USERNAME"
REPO_NAME       = "bankrisk-credit"

readme_content = f'''# 🏦 Credit Risk Analysis — BankRisk Intelligence Platform

![Python](https://img.shields.io/badge/python-3.10-blue)
![Pandas](https://img.shields.io/badge/pandas-2.x-orange)
![Scikit-learn](https://img.shields.io/badge/scikit--learn-latest-green)
![License](https://img.shields.io/badge/license-CC0-lightgrey)
![Status](https://img.shields.io/badge/status-en%20cours-yellow)

Analyse du risque de crédit sur le dataset Kaggle **Credit Risk Dataset**.  
Ce projet fait partie de la formation *BankRisk Intelligence Platform* et vise à construire un pipeline complet d’analyse, de scoring et d’explicabilité autour de la variable cible `loan_status`.

> **Taux de défaut de référence : 21,8 %**  
> 32 581 prêts · 12 variables · Cible binaire `loan_status` (0 = sain, 1 = défaut)

---

## 📖 Description

Ce projet a pour objectif d’explorer, nettoyer et modéliser un jeu de données de risque de crédit afin d’identifier les profils les plus risqués. Il couvre plusieurs étapes clés :

- exploration et analyse descriptive,
- ingénierie des features,
- modélisation supervisée,
- interprétabilité des prédictions,
- mise en place d’un workflow MLOps de base.

---

## 📁 Structure du repository

```text
{REPO_NAME}/
├── data/
│   ├── raw/            ← données brutes, non versionnées
│   └── processed/      ← données transformées et features
├── notebooks/          ← notebooks de formation et analyses
├── src/                ← scripts Python réutilisables
├── tests/              ← tests unitaires
├── requirements.txt    ← dépendances Python
└── .gitignore          ← exclusions Git
```

---

## ⚙️ Installation

```bash
git clone https://github.com/{GITHUB_USERNAME}/{REPO_NAME}.git
cd {REPO_NAME}
python -m venv .venv
source .venv/bin/activate  # macOS / Linux
# .venv\\Scripts\\activate  # Windows Git Bash
pip install -r requirements.txt
```

---

## 📊 Dataset

| Propriété | Valeur |
|-----------|--------|
| Source | Kaggle — Credit Risk Dataset |
| Licence | CC0 |
| Lignes | 32 581 |
| Colonnes | 12 |
| Cible | `loan_status` |
| Taux de défaut | **21,8 %** |

> Le fichier CSV doit être placé manuellement dans le dossier `data/raw/` avant l’exécution des notebooks.

---

## 🧰 Stack technique

- Python 3.10
- Pandas
- NumPy
- Plotly
- Scikit-learn
- MLflow
- DagsHub
- Hugging Face
- PySpark
- Delta Lake
- Streamlit
- GitHub Actions

---

## 🗓️ Roadmap sur 3 jours

| Jour | Session | Thème |
|------|---------|-------|
| J1 | S1 | Setup & GitHub Copilot |
| J1 | S2 | Pandas & NumPy |
| J1 | S3 | EDA & Plotly |
| J1 | S4 | Feature Engineering |
| J2 | S1 | K-Means Risk Profiling |
| J2 | S2 | Logistic Regression & Random Forest |
| J2 | S3 | MLflow & DagsHub |
| J2 | S4 | Explicabilité avec SHAP |
| J3 | S1 | Streamlit Dashboard |
| J3 | S2 | PySpark ETL |
| J3 | S3 | Streaming Structuré |
| J3 | S4 | Delta Lake & CI/CD |

---

## 📄 Licence

Le dataset est fourni sous licence **CC0**.  
Le code source de ce projet est proposé sous licence **MIT**.
'''

with open(ROOT / "README.md", "w", encoding="utf-8") as f:
    f.write(readme_content)

print("✅ README.md généré")
print(f"   Taille : {len(readme_content):,} caractères")
print("   Fichier écrit :", ROOT / "README.md")


### Étape 3 : Personnaliser avec Copilot `Ctrl+I`

| Sélectionner | Prompt Copilot (`Ctrl+I`) |
|---|---|
| La section description | `rends cette description plus percutante pour un recruteur bancaire` |
| Le tableau Stack | `ajoute une colonne 'version' avec les numéros de version actuels` |
| La section Roadmap | `transforme cette roadmap en tableau avec les dates et le statut` |
| L'ensemble du README | `traduis en anglais en gardant le même format` |


---

### 🖥️ Étape 4 : Premier commit depuis le terminal

Ouvrir le terminal intégré VS Code (`Ctrl+\``) et exécuter :

```bash
# Vérifier ce qui va être commité (les données ne doivent PAS apparaître)
git status

# Stager tous les fichiers (hors données — .gitignore fait son travail)
git add .

# Vérification : data/raw/*.csv ne doit PAS apparaître dans la liste
git status

# Committer
git commit -m "feat: J1S1 - structure + requirements + gitignore + README"

# Pousser vers GitHub
git push origin main
```

> **Résultat attendu :** sur `github.com/TON_USERNAME/bankrisk-credit`, la page d'accueil  
> affiche automatiquement le `README.md` rendu en HTML.


In [ ]:
# ── Vérification post-commit ─────────────────────────────────────────────────
import subprocess, os

os.chdir(ROOT)

result = subprocess.run("git log --oneline -5", shell=True, capture_output=True, text=True)
if result.returncode == 0:
    print("✅ Historique des 5 derniers commits :")
    print(result.stdout)
else:
    print("⚠️  Impossible de lire l'historique Git")
    print("   → Exécuter les commandes bash de l'Étape 4 dans le terminal")


---

## ✅ Checklist de fin de session J1S1

| | Étape |
|---|-------|
| ☐ | Python 3.10.14 · venv actif · kernel `.venv` sélectionné dans VS Code |
| ☐ | Tous les packages importables sans erreur (cellule Bloc 1) |
| ☐ | Structure créée : `data/raw/` · `data/processed/` · `notebooks/` · `src/` · `tests/` |
| ☐ | `.gitkeep` présents dans chaque dossier vide |
| ☐ | `credit_risk_dataset.csv` dans `data/raw/` |
| ☐ | `requirements.txt` généré à la racine |
| ☐ | `.gitignore` généré — `data/raw/*.csv` exclu, `.gitkeep` conservés |
| ☐ | Copilot actif — Exercice 1 : commentaire-intention → `Tab` fonctionne |
| ☐ | Exercice 2 : `Ctrl+I` → explication du `groupby` |
| ☐ | Exercice 3 : `Ctrl+I` → 3 erreurs corrigées |
| ☐ | Exercice 4 : `Ctrl+I` → code refactorisé |
| ☐ | Exercice 5 : chat latéral → `calculer_taux_defaut` générée et testée |
| ☐ | `df.shape` = `(32581, 12)` ✅ |
| ☐ | Taux de défaut **21,8 %** affiché ✅ |
| ☐ | `README.md` généré et personnalisé avec Copilot |
| ☐ | Premier commit pushé sur GitHub (`git log` affiche le commit) |

---

## ➡️ J1S2 — Pandas & NumPy

`describe()` · `groupby()` · `value_counts()` · types de données · export `data/processed/credit_features_j1.parquet`
